# Logistic Regression from Scratch

In this tutorial, you will learn
* What is logistic regression?
* What is log likelihood and maximum likelihood estimation?
* How can you implement logistic regression without using the scikit-learn fucntion.

## Introduction

Logistic regression is one of the most widely used machine learning algorithm for classification problems. It is a kind of generalised linear model (you can read more about generalised linear models [here](https://online.stat.psu.edu/stat504/node/216/)), just like the linear regression algorithm. The first question that would come to your mind is why is logistic regression needed when we already have linear regression. This is because linear regression is used to predict a continuous value. But in cases where we need a binary answer, that is, answer in 'Yes' or 'No', we use the logistic regression model.

To explain further, let's consider a popular example. We have a dataset of patients with different health parameters using which we have to predict which patient has heart disease and which patient doesn't. In this example we either want an output of 1 for 'Yes' or 0 for 'No'. A linear regression does not suffice in this case. This is where logistic regression comes into picture.

**Logistic Regression is a Machine Learning classification algorithm that is used to predict the probability of a categorical dependent variable.**

In simpler words, given a set of independent feature variables `X`, the logistic regression model gives us an output `y`, which is the probability of the given event occuring. In the above example, `y` would be the probability of our patient having a heart disease.

The output variable `y` of logistic regression is in the form of 'log(odds)'. Therefore, we must know about odds, probability and consequently `log(odds)` as well.

## Understanding Odds and Probability

***Odds*** is another way of measuring the probability of an event. Lets say the probability for an event is denoted by *P(E)*

Then the odds of the event will be = *P(E) / (1 - P(E))*
You might say that the odds and probability of an event are very similar measures. Why at all do we need both? This is because while probability is bounded between 0 to 1, odds can vary from 0 to infinity and log-odds vary between ngative infinity and positive infinity.

Lets revisit the equation for linear regression : 


$$Y = B0 + B*X$$

where,
    
   Y = Output variable
   
   B0 = y-axis intercept 
   
   B1 = set of coeeficients for feature variables X 
   
   X = set of feature variables in our dataset

As already mentioned earlier, the output of a logistic regression is log odds. So, the equation for logistic regression, though very similar to linear regression, has one major difference.
                    
$$Z = B0 + B*X$$

As you will notice, all terms are the same except Z. Z here denotes the log odds, which is our output.
But how do we find the probability of an event occuring from the log of odds? This is done by using the logit function, also called the sigmoid function.

$$Probability = 1 / (1 + e^(-Z))$$
                                        
The sigmoid curve has a finite limit of :

* 0 when x approaches negative infinity
* 1 when x approaches positive infinity

![sigmoid.png](attachment:sigmoid.png)

So far, we have understood the math behind logistic regression, but we still dont know the value of our coeeficients B0 and B1. In linear regression, we used the OLS (Ordinary Least Squares) method to find feature variable coefficients and the intercept. In logistic regression we use the method of Maximum Likelihood Estimation, also known as MLE. 
We will further see what is MLE and how can we implement it through code.

## Maximum Likelihood Estimation - Logistic Regression:

There can be infinite number of regression coefficients but the maximum likelihood estimate is that set of regression coefficients for which the probability of getting the data we have observed is maximum. 
It involves maximizing a likelihood function in order to find the probability distribution and parameters that best explain the observed data.

Here we are trying to find values of B {B0, B1, B2 ...} to fit our dataset X and then make an accurate prediction y. Thus we are trying the maximize our probability of observing a dataset X given a range of parameters B for our model.
This probability is represented as : *P(X | B)*

'|' denotes conditional probability, but commonly it is denoted by ';'. So *P(X | B)* = *P(X ; B)*.
This probability is called the likelihood function and is denoted by L(). Thus our likelihood function is **L(X ; B)**.
The aim of MLE is to maximize this function, that is, find *max(L(X ; B))*.

If we consider n samples in our dataset {x1, x2, x3..... xn} the likelihood function becomes: 
                                            
                                                   L({x1, x2, x3...xn ; B)
This can be unpacked as: 
                         
                                               product i to n L(xi ; B), or,
 $$\prod_{i = 1}^{n} L(x_i ; B)$$ 

As multplying so many probabilities together can be complex and unstable, we take log of the likelihood function. So after maximizing our expression changes to: 

                                                sum i to n logL(xi ; B) 
$$\sum_{i=1}^{n} logL(x_i ; B)$$                                            

In a logistic regression model, B is the parameters of our model. Using this model we will make our predictions from the given dataset. Let's take the example of supervised learning. We need to find the conditional probability of predicting output y given X (P(y | X)). The likelihood function changes to:

                                      maximize sum i to n logL(P(yi | xi) ; B), or,
$$\operatorname{argmax}\sum_{i=1}^{n} logL(x_i ; B)$$

To use the MLE function, we need a probability distribution of our output. The ouputs of a logistic regression can only have 2 results, 1 or 0. So we use a Bernoulli distribution (read more about Bernoulli distribution [here](https://towardsdatascience.com/understanding-bernoulli-and-binomial-distributions-a1eef4e0da8f)).
The Bernoulli distribution has a single parameter: the probability of a successful outcome (p).
* P(y=1) = p
* P(y=0) = 1 – p

The expected value (mean) of Bernoulli distribution is calculated as: 

                                            mean = p * 1 + (1 – p) * 0

This provides the basis for the likelihood function for a specific input, where the output probability is given by the model (yhat) and the actual label (y) is given from the dataset.

                                     likelihood = yhat * y + (1 – yhat) * (1 – y), or,
$$LL = y\hat{} * y + (1 – y\hat{}) * (1 – y)$$

This function will always return a large probability when the model is close to the matching class value, and a small value when it is far away, for both y=0 and y=1 cases.

A small implementation is shown below : 

In [0]:
# test of Bernoulli likelihood function

# likelihood function for Bernoulli distribution
def likelihood(y, yhat):
    return yhat * y + (1 - yhat) * (1 - y)

# test for y=1
y, yhat = 1, 0.9
print('y=%.1f, yhat=%.1f, likelihood: %.3f' % (y, yhat, likelihood(y, yhat)))
y, yhat = 1, 0.1
print('y=%.1f, yhat=%.1f, likelihood: %.3f' % (y, yhat, likelihood(y, yhat)))
# test for y=0
y, yhat = 0, 0.1
print('y=%.1f, yhat=%.1f, likelihood: %.3f' % (y, yhat, likelihood(y, yhat)))
y, yhat = 0, 0.9
print('y=%.1f, yhat=%.1f, likelihood: %.3f' % (y, yhat, likelihood(y, yhat)))

y=1.0, yhat=0.9, likelihood: 0.900
y=1.0, yhat=0.1, likelihood: 0.100
y=0.0, yhat=0.1, likelihood: 0.900
y=0.0, yhat=0.9, likelihood: 0.100


We can update the likelihood function using the log to transform it into a log-likelihood function:

                               log-likelihood = log(yhat) * y + log(1 – yhat) * (1 – y), or,
$$log(LL) = log(y\hat{}) * y + log(1 – y\hat{}) * (1 – y)$$
Finally, we can sum the likelihood function across all examples in the dataset to maximize the likelihood:

                        maximize sum i to n log(yhat_i) * yi + log(1 – yhat_i) * (1 – yi), or,
$$\operatorname{argmax}\sum_{i=1}^{n} log(y\hat{}_i) * y_i + log(1 – y\hat{}_i) * (1 – y_i)$$

## Estimating coefficients manually:

To estimate the coefficients of our model from scratch, you will first need a function that can predict the output probability of our model from `yhat`.
Below is a function named predict() that predicts an output value for a row given a set of coefficients.

The first coefficient is always the intercept, also called the bias or B0 as it is standalone and not responsible for a specific input value. The equation of our model in this case is the standard equation : 
$$Z = B0 + B1 * x1 + B2 * x2$$
where, B0 = intercept on y-axis and B1 & B2 = coefficient for X

In [0]:
# Make a prediction with coefficients
def predict(row, coefficients):
    yhat = coefficients[0] # This denotes the term B0
    for i in range(len(row)-1):
        yhat += coefficients[i + 1] * row[i] # This statement adds the term B1 * X
    return 1.0 / (1.0 + exp(-yhat)) # We return the probability using sigmoid function

To efficiently test our function and estimate coefficients you also need a create a dataset with values of X and y.
Below is a random dataset that will be further used in our implementation.

In [0]:
dataset = [[2.7810836,2.550537003,0],
    [1.465489372,2.362125076,0],
    [3.396561688,4.400293529,0],
    [1.38807019,1.850220317,0],
    [3.06407232,3.005305973,0],
    [7.627531214,2.759262235,1],
    [5.332441248,2.088626775,1],
    [6.922596716,1.77106367,1],
    [8.675418651,-0.242068655,1],
    [7.673756466,3.508563011,1]]

The dataset we have created above is an array with the following values for x1, x2 and y :

In [0]:
    X1        X2        Y
2.7810836  2.550537003  0
1.465489372 2.362125076 0
3.396561688 4.400293529 0
1.38807019  1.850220317 0
3.06407232  3.005305973 0
7.627531214 2.759262235 1
5.332441248 2.088626775 1
6.922596716 1.77106367  1
8.675418651 -0.24206865 1
7.673756466 3.508563011 1

The values of y are only 0 or 1, because the model we are trying to build is that of a logistic regression which is used for binary classification.

The next step is to determine the coefficients B0, B1 and B2 which is done using stochastic gradient descent (read further about gradient descent [here](https://machinelearningmastery.com/gradient-descent-for-machine-learning/).
You can calculate the new coefficient values using a simple update equation.

                      B{t) = B + alpha * (y(t) – yhat(t)) * yhat(t) * (1 – yhat(t)) * x, or,
$$ B(t) = B(t) + \alpha * (y – y\hat{}) * y\hat{} * (1 – y\hat{}) * x$$
The special coefficient at the beginning of the list, also called the intercept, is updated in a similar way, except without an input as it is not associated with a specific input value:

                      B0(t+1) = B0(t) + alpha * (y(t) - yhat(t)) * yhat(t) * (1 - yhat(t)), or,
$$B0(t+1) = B0(t) + \alpha * (y(t) - y\hat{}(t)) * y\hat{}(t) * (1 - y\hat{}(t))$$

Here, alpha is the learning rate of gradient descent and controls how much the coefficients (and therefore the model) changes or learns each time it is updated. Good values of alpha range between 0.1 and 0.3. For this example we will assume $$ \alpha = 0.1$$

Another parameter for carrying out gradient descent is number of epochs, which is he number of times to run through the training data while updating the coefficients.
We now have the learning rate alpha and the number of epochs. The function below will take in these parameters along with the dataset created above to give us a set of estimated coefficients.

In [0]:
# Estimate logistic regression coefficients using stochastic gradient descent
def coefficients_sgd(train, l_rate, n_epoch):
    coef = [0.0 for i in range(len(train[0]))]
    for epoch in range(n_epoch):
        for row in train:
            yhat = predict(row, coef) # You call the predict function here that was created above to get yhat
            error = row[-1] - yhat # the term of y-yhat
            coef[0] = coef[0] + l_rate * error * yhat * (1.0 - yhat) # Intercept B0 is estimated
            for i in range(len(row)-1):
                coef[i + 1] = coef[i + 1] + l_rate * error * yhat * (1.0 - yhat) * row[i] #B1 and B2 are estimated
    return coef

Now you can test this function on the created dataset. using `predict()` and `coefficients_sgd()`

In [0]:
from math import exp
 
# Make a prediction with coefficients
def predict(row, coefficients):
    yhat = coefficients[0]
    for i in range(len(row)-1):
        yhat += coefficients[i + 1] * row[i]
    return 1.0 / (1.0 + exp(-yhat))
 
# Estimate logistic regression coefficients using stochastic gradient descent
def coefficients_sgd(train, l_rate, n_epoch):
    coef = [0.0 for i in range(len(train[0]))]
    for epoch in range(n_epoch):
        for row in train:
            yhat = predict(row, coef)
            error = row[-1] - yhat
            coef[0] = coef[0] + l_rate * error * yhat * (1.0 - yhat)
            for i in range(len(row)-1):
                coef[i + 1] = coef[i + 1] + l_rate * error * yhat * (1.0 - yhat) * row[i]
    return coef
 
# Calculate coefficients
dataset = [[2.7810836,2.550537003,0],
          [1.465489372,2.362125076,0],
          [3.396561688,4.400293529,0],
          [1.38807019,1.850220317,0],
          [3.06407232,3.005305973,0],
          [7.627531214,2.759262235,1],
          [5.332441248,2.088626775,1],
          [6.922596716,1.77106367,1],
          [8.675418651,-0.242068655,1],
          [7.673756466,3.508563011,1]]
l_rate = 0.1
n_epoch = 30
coef = coefficients_sgd(dataset, l_rate, n_epoch)
print(coef)

[-0.4067126400924392, 0.8209365836324548, -1.1495205736938134]


As you can see above, the values for B0, B1 and B2 are respectively -0.4067126400924392, 0.8209365836324548, -1.1495205736938134. But to check if the model is giving the correct outputs, you can use the inbuilt `LogisticRegression()` model from scikit-learn and check the value of coeeficients. 
To do this you will have to divide your dataset into X and y, as shown below : 

In [0]:
X = [[2.7810836, 2.550537003],  
[1.465489372, 2.362125076], 
[3.396561688, 4.400293529], 
[1.38807019, 1.850220317], 
[3.06407232, 3.005305973], 
[7.627531214, 2.759262235], 
[5.332441248, 2.088626775], 
[6.922596716, 1.77106367],  
[8.675418651, -0.24206865], 
[7.673756466, 3.508563011]]

y = [0,0,0,0,0,1,1,1,1,1]

Now we can import the `LogisticRegression()` model and use it on the dataset to find B0, B1 and B2.

In [0]:
from sklearn.linear_model import LogisticRegression

clf = LogisticRegression(solver='liblinear') # Not specifying a solver will throw us a warning.
                                             # You can read more about the model's parameters from 
                                             # scikit-learn's documentation on logistic Regression
clf.fit(X, y)

In [0]:
print(clf.intercept_, clf.coef_)

[-0.43104649] [[ 0.80500337 -1.1224275 ]]


As you can see, The coefficients obtained from scikit-learn logistic regression model {-0.43104649, 0.80500337, -1.1224275} are very close to coefficients obtained from the function built by us to estimate B {-0.4067126400924392, 0.8209365836324548, -1.1495205736938134}. This proves the correcteness of our function.

#### References : 
1. Machine Learning Mastery [blog](https://machinelearningmastery.com/implement-logistic-regression-stochastic-gradient-descent-scratch-python/)
2. Towards Data Science [blog](https://towardsdatascience.com/real-world-implementation-of-logistic-regression-5136cefb8125)

                                                -------------